packages/server/src/procedure.ts

```diff
@@ -99,12 +100,25 @@ export abstract class Procedure<
     type,
     path,
   }: ProcedureCallOptions<TContext>): Promise<TOutput> {
-    for (const middleware of this.middlewares) {
-      await middleware({ ctx, type, path });
-    }
-    const input = this.parseInput(rawInput);
-    const output = await this.resolver({ ctx, input, type });
-    return output;
+    const opts = { ctx, type, path, input: this.parseInput(rawInput) };
```

Comment:
I'm unsure about this --

Previously, middlewares had knowledge about `input` and we wouldn't even _get to_ the input parser unless all the middlewares had been run; was this an intentional change of yours or an unintentional side-effect? 

I was intentional with the previous behavior which made sure that nothing in the procedures were called if all middlewares didn't run first.

Reply:
It was intentional, or at least it was needed for `opts` to be passed to `this.resolver`. I like it though, partly for consistency - the `this.resolver` function expects an `input` prop, and is now effectively acting like a middleware. It's a little confusing for middlewares to receive everything _except_ input. But also it makes sense to me for parsing to happen first. Middleware could have side effects and it's helpful for trpc to filter out bad requests before they get to the middleware.

Example use case: a [fhir server](https://www.hl7.org/fhir) of some kind. On a previous team we used codegen to create io-ts codecs for every fhir resource. Something like that could be extended and plugged into trpc, and a middleware could log statistics per fhir resource type. For that to happen we'd need either `input` or `rawInput`, but it's more useful to pass `input`, especially since the parsing step might do some hydration of iso strings to dates, etc.

>I was intentional with the previous behavior which made sure that nothing in the procedures were called if all middlewares didn't run first.

In the new model, where middlewares are enclosing rather than preceding the procedure part, does that still apply?

I'm struggling to think of a use case where I'd want the middleware to still run on invalid input (in the world of http status codes for a regular server, I'd probably expect a malformed request which also had an invalid token to give me a 400, not a 401 or 403).

Reply:
Thanks for spending so much time thinking on this! 🙏 

Why I think it should be the other way around:

- Intuitive: a middleware is declared before a resolver, hence, should always be called first
- The input parser could later be turned into a potentially expensive async method as well (think admin routes)
- The input parser could leak information about the internals of a protected route (again, think admin routes)

---

If you think we need to have `rawInput` (or `input`) in middlewares we can add that at a later stage when we need it; it's much easier to add code than to delete it. Potentially it could be returned from the `next` function.

---

Another thought - rather than dealing with `Promise` and `try/catch`, how about we return a discriminated union similar to [zod's `safeParse`](https://github.com/colinhacks/zod#safeparse)?

```ts
router
  .middleware(async ({ ctx, path, type, next }) => {
      const result = await next();
     if (!result.ok && result.error.originalError instanceof RecoverableError) {
        return result.e.recover();
     }
     return result;
  })
```
The interface could look like this

```
import { ProcedureType } from '../router';
import { TRPCError } from '../TRPCError';

export const middlewareMarker = Symbol('middlewareMarker');
interface MiddlewareResultBase {
  /**
   * All middlewares should pass through their `next()`'s output.
   * Requiring this marker makes sure that can't be forgotten at compile-time.
   */
  readonly marker: typeof middlewareMarker;
}

interface MiddlewareOKResult extends MiddlewareResultBase {
  ok: true;
  data: unknown;
  // this could be extended with `input`/`rawInput` later
}
interface MiddlewareErrorResult extends MiddlewareResultBase {
  ok: false;
  error: TRPCError; // we could guarantee it's always of this type
  // this could be extended with `input`/`rawInput` later
}

type MiddlewareResult = MiddlewareOKResult | MiddlewareErrorResult;

export type MiddlewareFunction<TContext> = (opts: {
  ctx: TContext;
  type: ProcedureType;
  path: string;
  next: () => Promise<MiddlewareResult>;
}) => Promise<MiddlewareResult>;
```

Reply:
See 31efba0 suggestion

Reply:
>If you think we need to have rawInput (or input) in middlewares we can add that at a later stage when we need it

👍👍👍 Very good point, I'm absolutely ok with leaving out for now and possibly forever.

___

Re the discriminated union - will reply inline.